src/go/rpk/pkg/cli/cmd/wasm/common.go

```diff
@@ -1,133 +1,90 @@
-			Value: []byte("deploy"),
-		}, {
-			Key:   []byte("description"),
-			Value: []byte(description),
-		}, {
-			Key:   []byte("sha256"),
-			Value: shaValue[:],
-		}, {
-			Key:   []byte("type"),
-			Value: []byte(coprocType),
-		},
-	}
-	id := xxhash.Sum64([]byte(name))
-	binaryId := make([]byte, 8)
-	binary.LittleEndian.PutUint64(binaryId, id)
-	return sarama.ProducerMessage{
-		Key:     sarama.ByteEncoder(binaryId),
-		Topic:   kafka.CoprocessorTopic,
-		Value:   sarama.ByteEncoder(content),
-		Headers: headers,
-	}
+func nameIDKey(name string) []byte {
+	hash := xxhash.Sum64([]byte(name))
+	id := make([]byte, 8)
+	binary.LittleEndian.PutUint64(id, hash)
+	return id
 }
 
-func CreateRemoveMsg(name string, coprocType string) sarama.ProducerMessage {
-	var headers = []sarama.RecordHeader{
-		{
-			Key:   []byte("action"),
-			Value: []byte("remove"),
-		}, {
-			Key:   []byte("type"),
-			Value: []byte(coprocType),
-		},
+// We always try creating the coproc topic to ensure it exists, creating a
+// topic that already exists is fine. We use replicas -1 to ensure we use
+// the broker default, which is now three for three+ node clusters.
+func ensureCoprocTopic(cl *kgo.Client) error {
+	req := kmsg.NewCreateTopicsRequest()
+	t := kmsg.NewCreateTopicsRequestTopic()
+	t.Topic = coprocTopic
+	t.NumPartitions = 1
+	t.ReplicationFactor = -1
+	for _, kvs := range [][2]string{
+		{"cleanup.policy", "compact"},
+		{"compression.type", "zstd"},
+	} {
+		c := kmsg.NewCreateTopicsRequestTopicConfig()
+		c.Name = kvs[0]
+		c.Value = kmsg.StringPtr(kvs[1])
+		t.Configs = append(t.Configs, c)
 	}
-	id := xxhash.Sum64([]byte(name))
-	binaryId := make([]byte, 8)
-	binary.LittleEndian.PutUint64(binaryId, id)
-	return sarama.ProducerMessage{
-		Key:   sarama.ByteEncoder(binaryId),
-		Topic: kafka.CoprocessorTopic,
-		// create empty message, the remove command doesn't need
-		// information on message, just a key value
-		Value:   sarama.ByteEncoder([]byte{}),
-		Headers: headers,
+	req.Topics = append(req.Topics, t)
+	resp, err := req.RequestWith(context.Background(), cl)
+	if err != nil {
+		return err
 	}
+	err = kerr.ErrorForCode(resp.Topics[0].ErrorCode)
+	switch err {
+	case nil:
+		// topic did not exist
+	case kerr.TopicAlreadyExists:
+		// topic exists, fine
+	default:
+		return err
+	}
+	return nil
 }
 
-func AddTypeFlag(command *cobra.Command, coprocType *string) {
-	command.Flags().StringVar(
-		coprocType,
-		"type",
-		"async",
-		"WASM engine type (async, data-policy)",
-	)
-}
-
-func CheckCoprocType(coprocType string) error {
-	switch coprocType {
-	case "async", "data-policy":
-		return nil
-	default:
-		return fmt.Errorf("Unexpected WASM engine type: '%s'", coprocType)
+func produce(
+	cl *kgo.Client, name, description, coprocType string, file []byte,
+) error {
+	rec := kgo.KeySliceRecord(nameIDKey(name), nil)
+	rec.Topic = coprocTopic
+	rec.Headers = []kgo.RecordHeader{
+		{Key: "action", Value: []byte("remove")},
```

Comment:
Is this the right action here? I will request the coproc contributors because I'm not too familiar with the inner workings of coproc.

Reply:
Should be -- this is a refactoring of `CreateRemoveMessage` and `CreateDeployMessage`:
https://github.com/vectorizedio/redpanda/pull/2432/files/70d6fce362c8085acefee9f987a7bb550f3a422b..998237a58f926ebbddb1106693ca83817943c3ea#diff-89dcdb89fabb6cb1f3743b0493af1234dfcaf60917aa2d1de62ec9c8da44c12fL98

Reply:
Ah, I just saw that if `file != nil`, `rec.Headers` gets overwritten. Thanks!